instance/models/utils.py

```diff
@@ -44,3 +55,263 @@ def save(self, *args, **kwargs):
+                """
+                state = self.__get__(resource)
+                if not isinstance(state, accepted_states):
+                    raise WrongStateException("The method '{}' cannot be called in this state ({} / {}).".format(
+                        method.__name__, state.name, state.__class__.__name__  # pylint: disable=no-member
+                    ))
+
+            descriptor = self
+
+            class MagicWrapper:
+                """ Class which can wrap a method; the result can be used as a property or as a method. """
+                def __call__(self, resource):
+                    """ We are wrapping a property, not a method. No fancy stuff needed. """
+                    require_valid_state(resource)
+                    return method(resource)
+
+                def __get__(self, resource, _type):
+                    """ Get the (wrapped) method, and add a .is_available method to it """
+                    def wrapped_method(*args, **kwargs):
+                        """ Wrapper around the method which checks the state requirements first """
+                        require_valid_state(resource)
+                        return method(resource, *args, **kwargs)
+                    wrapped_method.is_available = lambda: isinstance(descriptor.__get__(resource), accepted_states)
+                    return wrapped_method
+            return MagicWrapper()
+        return wrap
+
+    def transition(self, to_state, from_states=None):
+        """
+        Returns a "transition" method, which can be called to change the current state from
+        state(s) 'from' to 'to'.
+
+        from_states can be a state class, a tuple of classes, a parent class or mixin of various
+        state classes, a tuple of mixins, etc.. if None, any valid state will be accepted.
+        """
+        assert to_state in self.state_classes
+
+        def do_transition(resource):
+            """ The method that performs a specific transition """
+            current_state = self.__get__(resource)
+            if from_states and not isinstance(current_state, from_states):
+                raise WrongStateException("This transition cannot be used to move from {} to {}".format(
+                    current_state.name, to_state.name  # pylint: disable=no-member
+                ))
+            self._set_state(resource, to_state)
+        do_transition.from_states = from_states  # Convenient way for other code to inspect this transition
+        do_transition.to_state = to_state  # Convenient way for other code to inspect this transition
+        return do_transition
+
+    @property
+    def model_field_choices(self):
+        """
+        Get a tuple of (state_id, name) pairs.
+
+        Suitable for passing to a django CharField choices parameter.
+        """
+        return tuple((state.state_id, state.__class__.__name__) for state in self.state_classes)
+
+    # Internal helper methods:
+
+    def _get_initial_state(self, resource):
```

Comment:
@bradenmacdonald nit: a bit confusing name - it clashes with "initial_state" of the actual state-machine-backed property (that becomes `default_state_class` here). Actually, it tries fetching *current* value from DB, or falls back to default if it's not in DB yet. `_load_state(self, resource, default=self.default_state)` maybe?